lib/cli/kit/command_registry.rb

```diff
@@ -1,71 +1,145 @@
+# typed: true
 require 'cli/kit'
 
 module CLI
   module Kit
     class CommandRegistry
+      extend(T::Sig)
+
       attr_reader :commands, :aliases
 
+      module ContextualResolverInterface
+        extend(T::Sig)
+        extend(T::Helpers)
+        interface!
+
+        sig { abstract.returns(T::Array[String]) }
+        def command_names; end
+
+        sig { abstract.returns(T::Hash[String, String]) }
+        def aliases; end
+
+        sig { abstract.params(name: String).returns(T.nilable(Class)) }
+        def command_class(name); end
+      end
+
       module NullContextualResolver
+        extend(T::Sig)
+        extend(ContextualResolverInterface)
+
+        sig { implementation.returns(T::Array[String]) }
         def self.command_names
           []
         end
 
+        sig { implementation.returns(T::Hash[String, String]) }
         def self.aliases
           {}
         end
 
+        sig { implementation.params(_name: String).returns(T.nilable(Class)) }
         def self.command_class(_name)
           nil
         end
       end
 
+      sig do
+        params(
+          default: T.nilable(String),
+          contextual_resolver: T.untyped,
+        ).void
+      end
       def initialize(default:, contextual_resolver: nil)
         @commands = {}
         @aliases  = {}
         @default = default
         @contextual_resolver = contextual_resolver || NullContextualResolver
       end
 
+      sig { returns(T::Hash[String, String]) }
       def resolved_commands
         @commands.each_with_object({}) do |(k, v), a|
           a[k] = resolve_class(v)
         end
       end
 
+      sig do
+        params(
+          const: Symbol, # constant name that should be loaded later upon use
+          name: String,  # name of command that will be used to key lookups
+        ).void
+      end
       def add(const, name)
         commands[name] = const
       end
 
+      sig do
+        params(name: String) # candidate command name
+          .returns([
```

Comment:
Why using a tuple and not a concrete class here?

Reply:
Indeed, that would be a better design, but it would impact the public API in a way that would matter.